pkg/inspection/backup.go

```diff
@@ -0,0 +1,135 @@
+ *
+ * Copyright (C) 2012-2023 Tencent. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package inspection
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	tencentCOS "github.com/tencentyun/cos-go-sdk-v5"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+
+	backupapiv2 "github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta2"
+	kstonev1alpha1 "tkestack.io/kstone/pkg/apis/kstone/v1alpha1"
+	"tkestack.io/kstone/pkg/backup"
+	// import backup provider
+	_ "tkestack.io/kstone/pkg/backup/providers"
+	"tkestack.io/kstone/pkg/inspection/metrics"
+)
+
+// CollectionBackupfileSuccess checks backup succeed or not  , and
+// transfer it to prometheus metrics
+func (c *Server) CollectionBackupfileSuccess(inspection *kstonev1alpha1.EtcdInspection) error {
+	namespace, name := inspection.Namespace, inspection.Spec.ClusterName
+	labels := map[string]string{
+		"clusterName": name,
+	}
+	cluster, err := c.cli.KstoneV1alpha1().EtcdClusters(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("cluster '%s' in work queue no longer exists", name))
+			return nil
+		}
+		metrics.BackupSucceed.With(labels).Set(0)
+		return err
+	}
+
+	// generate backup config
+	strCfg, found := cluster.Annotations[backup.AnnoBackupConfig]
+	if !found || strCfg == "" {
+		err = fmt.Errorf(
+			"backup config not found, annotation key %s not exists, namespace is %s, name is %s",
+			backup.AnnoBackupConfig,
+			cluster.Namespace,
+			cluster.Name,
+		)
+		klog.Errorf(err.Error())
+		return err
+	}
+	backupConfig := &backup.Config{}
+	err = json.Unmarshal([]byte(strCfg), backupConfig)
+	if err != nil {
+		klog.Errorf(err.Error())
+		metrics.BackupSucceed.With(labels).Set(0)
+		return err
+	}
+
+	// generate backup provider
+	backupProvider, err := backup.GetBackupProvider(string(backupConfig.StorageType), &backup.ProviderConfig{
+		Kubeconfig: "",
+	})
+	if err != nil {
+		klog.Errorf(err.Error())
+		return err
+	}
+	resp, err := backupProvider.List(cluster)
+	if err != nil {
+		klog.Errorf(err.Error())
+		metrics.BackupSucceed.With(labels).Set(0)
+		return err
+	}
+	switch backupConfig.StorageType {
+	case backupapiv2.BackupStorageTypeCOS:
+		err = checkBackupCOS(resp, backupConfig.StoragePolicy.BackupIntervalInSecond)
```

Comment:
checkBackupCOS -> StatBackupFiles, It is recommended to add an abstract interface in the backup provider.

Reply:
https://github.com/maudL1n/kstone/blob/feature-checkbackupfile/pkg/backup/plugins.go